File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, h: List[int]) -> int:
        """
        
        0 1 2 0 3 4 5

        """

        res = 0
        l, r = 0, len(h)-1

        

        lMax, rMax = 0, 0

        while l <= r:
            if h[l] <= h[r]:
                if h[l] < lMax and lMax > 0:
                    res += lMax - h[l]
                lMax = max(h[l], lMax)
                l += 1
            else:
                if h[r] < rMax and rMax > 0:
                    res += rMax - h[r]
                rMax = max(h[r], rMax)
                r -= 1

        return res

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] g) {
        int n = g.length, l = 0, r = n-1, res = 0, 
                lMax = 0, rMax= 0;
        while (l <= r) {
            if (lMax <= rMax) {
                if (g[l] > lMax) {
                    lMax = g[l++];
                } else{
                    res += (lMax - g[l++]);
                }
            } else  {
                if (g[r] > rMax) {
                    rMax = g[r--];
                } else{
                    res += (rMax - g[r--]);
                }
            }
        }
        return res;
    }
}